doc/crdt-details.md

```diff
@@ -0,0 +1,1054 @@
+                if undone_groups.contains(undo_group) {
+                    // no need to un-delete undone inserts since we'll just shrink them out
+                    Cow::Owned(deletes_from_union.transform_shrink(inserts))
+                } else {
+                    let un_deleted = deletes_from_union.subtract(deletes);
+                    Cow::Owned(un_deleted.transform_shrink(inserts))
+                }
+            }
+            Undo { ref toggled_groups, ref deletes_bitxor } => {
+                if invert_undos {
+                    let new_undone = undone_groups.symmetric_difference(toggled_groups).cloned().collect();
+                    undone_groups = Cow::Owned(new_undone);
+                    Cow::Owned(deletes_from_union.bitxor(deletes_bitxor))
+                } else {
+                    deletes_from_union
+                }
+            }
+        }
+    }
+    deletes_from_union
+}
+```
+
+#### Delta::synthesize
+
+This function is just a big hand-written triply-nested looping algorithm that loops over a `from_dels` and `to_dels` subset and uses a `tombstones` to construct a `Delta`. It's not particularly interesting, easy or worthwhile to understand.
+
+```rust
+/// Synthesize a delta from a "union string" and two subsets: an old set
+/// of deletions and a new set of deletions from the union. The Delta is
+/// from text to text, not union to union; anything in both subsets will
+/// be assumed to be missing from the Delta base and the new text. You can
+/// also think of these as a set of insertions and one of deletions, with
+/// overlap doing nothing. This is basically the inverse of `factor`.
+///
+/// Since only the deleted portions of the union string are necessary,
+/// instead of requiring a union string the function takes a `tombstones`
+/// rope which contains the deleted portions of the union string. The
+/// `from_dels` subset must be the interleaving of `tombstones` into the
+/// union string.
+pub fn synthesize(tombstones: &Node<N>, from_dels: &Subset, to_dels: &Subset) -> Delta<N> {
+  // ...
+}
+```
+
+### Engine::delta_rev_head
+
+This operation is similar to `Engine::get_head` except it returns a `Delta` from the text at a specified revision to the current head text. This is useful for things like updating the position of cursors and rich text spans when edits are made.
+
+Like `Engine::get_head` it starts by calling `Engine::deletes_from_cur_union_for_index` to get a `Subset` describing the state of the text at the old revision relative to the current union string. Now we can just use `Delta::synthesize` to create a `Delta` from the old to the new `deletes_from_union`. The problem is, `Delta::synthesize` expects the tombstones `Rope` you give it to correspond to `from_dels`, but we have one for `to_dels`. To fix this, we can use a helper called `shuffle_tombstones` to shuffle things to get an `old_tombstones` corresponding to `from_dels`.
+
+#### shuffle_tombstones
```

Comment:
General naming nit: I do not find it especially helpful that there are a bunch of functions with "shuffle" in their name that are not only _not_ doing any shuffling, but are actually quite opinionated about orderings. 😁